src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3978,14 +4002,214 @@ bool   os::win32::_is_windows_server         = false;
+            warning("Could not determine processor count from the job object.");
+            assert(false, "Must find at least 1 logical processor");
+          }
+        } else {
+          char buf[512];
+          size_t buf_len = os::lasterror(buf, sizeof(buf));
+          warning("Attempt to query job object information failed: %s", buf_len != 0 ? buf : "<unknown error>");
+        }
+
+        os::free(job_object_information);
+    } else {
+        warning("os::malloc() failed to allocate %ld bytes for QueryInformationJobObject", job_object_information_length);
+    }
+  } else {
+    char buf[512];
+    size_t buf_len = os::lasterror(buf, sizeof(buf));
+    warning("Attempt to query job object information failed: %s", buf_len != 0 ? buf : "<unknown error>");
+    assert(false, "Unexpected QueryInformationJobObject error code");
+    return 0;
+  }
+
+  return processors;
+}
+
+DWORD os::win32::system_logical_processor_count() {
+  DWORD logical_processors = 0;
+  LOGICAL_PROCESSOR_RELATIONSHIP relationship_type = RelationGroup;
+  PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX system_logical_processor_info = nullptr;
+  DWORD returned_length = 0;
+
+  if (GetLogicalProcessorInformationEx(relationship_type, nullptr, &returned_length) == 0) {
+    DWORD last_error = GetLastError();
+
+    if (last_error == ERROR_INSUFFICIENT_BUFFER) {
+      system_logical_processor_info = (PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX)os::malloc(returned_length, mtInternal);
+
+      if (nullptr == system_logical_processor_info) {
+        warning("os::malloc() failed to allocate %ld bytes for GetLogicalProcessorInformationEx buffer", returned_length);
+      } else if (GetLogicalProcessorInformationEx(relationship_type, system_logical_processor_info, &returned_length) == 0) {
+        char buf[512];
+        size_t buf_len = os::lasterror(buf, sizeof(buf));
+        warning("Attempt to determine logical processor count from GetLogicalProcessorInformationEx() failed: %s", buf_len != 0 ? buf : "<unknown error>");
+      } else {
+        DWORD processor_groups = system_logical_processor_info->Group.ActiveGroupCount;
+
+        PROCESSOR_GROUP_INFO* group_info = (PROCESSOR_GROUP_INFO*)system_logical_processor_info->Group.GroupInfo;
+        for (DWORD i = 0; i < processor_groups; i++, group_info++) {
+          logical_processors += group_info->ActiveProcessorCount;
+        }
+
+        if (logical_processors == 0) {
+          warning("Could not determine logical processor count from GetLogicalProcessorInformationEx()");
+          assert(false, "Must find at least 1 logical processor");
+        }
+      }
+
+      os::free(system_logical_processor_info);
+    } else {
+      char buf[512];
+      size_t buf_len = os::lasterror(buf, sizeof(buf));
+      warning("Attempt to determine logical processor count from GetLogicalProcessorInformationEx() failed: %s", buf_len != 0 ? buf : "<unknown error>");
+    }
+  }
+
+  return logical_processors;
```

Comment:
Before returning please add a `log_debug(os)` statement indicating the number of logical processors discovered.

Reply:
This message will not appear in the log if the logging statement is placed here because the [log configuration is initialized](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/src/hotspot/share/runtime/threads.cpp#L449) after [initializing the os module](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/src/hotspot/share/runtime/threads.cpp#L431). I have added logging code that will be executed only the first time the active processor count is retrieved.